pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go

```diff
@@ -39,28 +43,37 @@ const (
 	// Name is the name of the plugin used in the plugin registry and configurations.
 	Name = "NodeAffinity"
 
-	// ErrReason for node affinity/selector not matching.
-	ErrReason = "node(s) didn't match node selector"
+	// ErrReasonPod is the reason for Pod's node affinity/selector not matching.
+	ErrReasonPod = "node(s) didn't match Pod's node affinity"
+
+	// errReasonEnforced is the reason for added node affinity not matching.
+	errReasonEnforced = "node(s) didn't match scheduler-enforced node affinity"
 )
 
 // Name returns name of the plugin. It is used in logs, etc.
 func (pl *NodeAffinity) Name() string {
 	return Name
 }
 
-// Filter invoked at the filter extension point.
+// Filter checks if the Node matches the Pod .spec.affinity.nodeAffinity and
+// the plugin's added affinity.
 func (pl *NodeAffinity) Filter(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
 	node := nodeInfo.Node()
 	if node == nil {
 		return framework.NewStatus(framework.Error, "node not found")
 	}
+	if pl.addedNodeSelector != nil && !pl.addedNodeSelector.Match(node) {
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, errReasonEnforced)
+	}
 	if !pluginhelper.PodMatchesNodeSelectorAndAffinityTerms(pod, node) {
-		return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReason)
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonPod)
 	}
 	return nil
```

Comment:
Nit: is it more effient to do a "smart merge" on addedAffinity and user-specified ones? so that we only run "match logic" once. For example, "foo in [bar]" and  "foo in [bar]" gets de-duplicated to "foo in [bar]"; while "foo in [bar]" and  "foo in [baz]" gets merged to one entry: "foo in [bar, baz]".

Reply:
I don't think it's worth.
See my documentation PR https://github.com/kubernetes/website/pull/24914/files#diff-ba98b126fe5dd6bf0570b39d4928eb9143d1f2ed60fbaf9fa2051810acc2e8a6R192

Reply:
SG.